File: backend/bin/src/controllers/prescription.controller.js
const { Prescription, Appointment, Doctor, Patient, User } = require('../models');
const { notifyUser } = require('../services/socket.service');

const createPrescription = async (req, res) => {
  try {
    const { appointment_id, medicines, notes } = req.body;

    if (!appointment_id || !medicines) {
      return res.status(400).json({ success: false, message: 'Appointment ID and medicines list are required.' });
    }

    // 1. Verify doctor authorization for this appointment
    const doctor = await Doctor.findOne({ where: { user_id: req.user.id } });
    if (!doctor) {
      return res.status(403).json({ success: false, message: 'Only registered doctors can create prescriptions.' });
    }

    const appointment = await Appointment.findByPk(appointment_id, {
      include: [{ model: Patient, include: [{ model: User, attributes: ['id', 'name'] }] }]
    });

    if (!appointment) {
      return res.status(404).json({ success: false, message: 'Appointment not found.' });
    }

    if (appointment.doctor_id !== doctor.doctor_id) {
      return res.status(403).json({ success: false, message: 'You are not authorized to prescribe for this appointment.' });
    }

    // 2. Create or Update Prescription
    let prescription = await Prescription.findOne({ where: { appointment_id } });

    if (prescription) {
      prescription.medicines = medicines;
      prescription.notes = notes;
      await prescription.save();
    } else {
      prescription = await Prescription.create({
        appointment_id,
        medicines,
        notes
      });
    }

    // 3. Mark appointment as completed when prescription is uploaded
    appointment.status = 'completed';
    await appointment.save();

    // 4. Notify patient
    const io = req.app.get('io');
    notifyUser(io, appointment.Patient.User.id, {
      title: 'New Prescription Uploaded',
      message: `Doctor has uploaded your prescription details for appointment on ${appointment.date}.`
    });

    res.status(201).json({
      success: true,
      message: 'Prescription processed successfully and appointment marked as completed.',
      data: prescription
    });
  } catch (error) {
    console.error('Create Prescription Error:', error);
    res.status(500).json({ success: false, message: 'Failed to process prescription.' });
  }
};

const getPrescription = async (req, res) => {
  try {
    const { appointmentId } = req.params;

    const prescription = await Prescription.findOne({
      where: { appointment_id: appointmentId },
      include: [
        {
          model: Appointment,
          include: [
            { model: Patient, include: [{ model: User, attributes: ['name'] }] },
            { model: Doctor, include: [{ model: User, attributes: ['name'] }] }
          ]
        }
      ]
    });

    if (!prescription) {
      return res.status(404).json({ success: false, message: 'Prescription not found.' });
    }

    // Ensure requesting user is authorized (is the patient, the doctor, or an admin)
    const userId = req.user.id;
    const isPatientUser = prescription.Appointment.Patient.user_id === userId;
    const isDoctorUser = prescription.Appointment.Doctor.user_id === userId;
    const isAdminUser = req.user.role === 'admin';

    if (!isPatientUser && !isDoctorUser && !isAdminUser) {
      return res.status(403).json({ success: false, message: 'Not authorized to view this prescription.' });
    }

    res.status(200).json({ success: true, data: prescription });
  } catch (error) {
    console.error('Get Prescription Error:', error);
    res.status(500).json({ success: false, message: 'Failed to retrieve prescription.' });
  }
};

module.exports = {
  createPrescription,
  getPrescription
};

File: backend/src/main/java/com/telehealth/pro/controller/PrescriptionController.java
package com.telehealth.pro.controller;

import com.telehealth.pro.dto.ApiResponse;
import com.telehealth.pro.entity.*;
import com.telehealth.pro.repository.*;
import com.telehealth.pro.security.CustomUserDetails;
import com.telehealth.pro.service.SocketService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.bind.annotation.*;

import java.util.Map;
import java.util.Optional;

@RestController
@RequestMapping("/api/prescriptions")
public class PrescriptionController {

    @Autowired
    private PrescriptionRepository prescriptionRepository;

    @Autowired
    private AppointmentRepository appointmentRepository;

    @Autowired
    private DoctorRepository doctorRepository;

    @Autowired
    private NotificationRepository notificationRepository;

    @Autowired
    private SocketService socketService;

    @PostMapping
    @PreAuthorize("hasRole('DOCTOR')")
    @Transactional
    public ResponseEntity<?> createPrescription(
            @AuthenticationPrincipal CustomUserDetails userDetails,
            @RequestBody Map<String, Object> body) {
        
        User currentUser = userDetails.getUser();

        if (body.get("appointment_id") == null || body.get("medicines") == null) {
            return ResponseEntity.status(HttpStatus.BAD_REQUEST)
                    .body(ApiResponse.error("Appointment ID and medicines list are required."));
        }

        Long appointmentId = ((Number) body.get("appointment_id")).longValue();
        String medicines = (String) body.get("medicines");
        String notes = (String) body.get("notes");

        Optional<Doctor> doctorOpt = doctorRepository.findByUser_Id(currentUser.getId());
        if (doctorOpt.isEmpty()) {
            return ResponseEntity.status(HttpStatus.FORBIDDEN)
                    .body(ApiResponse.error("Only registered doctors can create prescriptions."));
        }
        Doctor doctor = doctorOpt.get();

        Optional<Appointment> appointmentOpt = appointmentRepository.findById(appointmentId);
        if (appointmentOpt.isEmpty()) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                    .body(ApiResponse.error("Appointment not found."));
        }
        Appointment appointment = appointmentOpt.get();

        if (!appointment.getDoctor().getDoctorId().equals(doctor.getDoctorId())) {
            return ResponseEntity.status(HttpStatus.FORBIDDEN)
                    .body(ApiResponse.error("You are not authorized to prescribe for this appointment."));
        }

        Optional<Prescription> prescriptionOpt = prescriptionRepository.findByAppointmentAppointmentId(appointmentId);
        Prescription prescription;

        if (prescriptionOpt.isPresent()) {
            prescription = prescriptionOpt.get();
            prescription.setMedicines(medicines);
            prescription.setNotes(notes);
            prescription = prescriptionRepository.save(prescription);
        } else {
            prescription = Prescription.builder()
                    .appointment(appointment)
                    .medicines(medicines)
                    .notes(notes)
                    .build();
            prescription = prescriptionRepository.save(prescription);
        }

        // Complete the appointment
        appointment.setStatus("completed");
        appointmentRepository.save(appointment);

        // Notify Patient
        String alertTitle = "New Prescription Uploaded";
        String alertMsg = "Doctor has uploaded your prescription details for appointment on " + appointment.getDate() + ".";

        Notification notification = Notification.builder()
                .user(appointment.getPatient().getUser())
                .title(alertTitle)
                .message(alertMsg)
                .build();
        notification = notificationRepository.save(notification);

        // Emit live socket event
        socketService.notifyUser(appointment.getPatient().getUser().getId(), notification);

        return ResponseEntity.status(HttpStatus.CREATED)
                .body(ApiResponse.success("Prescription processed successfully and appointment marked as completed.", prescription));
    }

    @GetMapping("/appointment/{appointmentId}")
    @PreAuthorize("isAuthenticated()")
    public ResponseEntity<?> getPrescription(
            @AuthenticationPrincipal CustomUserDetails userDetails,
            @PathVariable Long appointmentId) {
        
        User currentUser = userDetails.getUser();
        Optional<Prescription> prescriptionOpt = prescriptionRepository.findByAppointmentAppointmentId(appointmentId);
        if (prescriptionOpt.isEmpty()) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                    .body(ApiResponse.error("Prescription not found."));
        }

        Prescription prescription = prescriptionOpt.get();

        boolean isPatientUser = prescription.getAppointment().getPatient().getUser().getId().equals(currentUser.getId());
        boolean isDoctorUser = prescription.getAppointment().getDoctor().getUser().getId().equals(currentUser.getId());
        boolean isAdminUser = "admin".equals(currentUser.getRole());

        if (!isPatientUser && !isDoctorUser && !isAdminUser) {
            return ResponseEntity.status(HttpStatus.FORBIDDEN)
                    .body(ApiResponse.error("Not authorized to view this prescription."));
        }

        return ResponseEntity.ok(ApiResponse.success(prescription));
    }
}
